### ISSUE TITLE
Add custom migration generation hooks

### ISSUE THREAD

**Issue by Ginden** [typeorm/typeorm#8544]
## Feature Description

Allow users to extend migration generation.

### The Problem

We have built custom trigger system, because our database is accessed by multiple services written in languages other than Node. This system is quite error prone.

Potentially, this system allow user to specify:

* Triggers
* [Rules](https://www.postgresql.org/docs/14/sql-createrule.html)
* Stored procedures
* Functions

### The Solution

Allow to specify option `queryBuilderHooks` in `ormconfig` and connection hooks with type `{queryBuilderHooks: (typeof QueryBuilderHook)[]`.

Following interfaces:

```ts
abstract class QueryBuilderHook {
   init?(queryRunner: QueryRunner, schemaBuilder: SchemaBuilder): Promise<void>;
   afterAll?(queryRunner: QueryRunner, schemaBuilder: SchemaBuilder): Promise<SqlInMemory>;
   beforeAll?(queryRunner: QueryRunner, schemaBuilder: SchemaBuilder): Promise<SqlInMemory>;
}
```

#### Possible enhacements

Let's keep it simple for a moment.

* All `RdbmsSchemaBuilder` methods can have `before${method}` and `after${method}`
* Hypothetical package `@typeorm/postgres` can have support for Postgres-specific extensions, like rules. 

### Considered Alternatives

We already have trigger system.

### Relevant Database Driver(s)

| DB Type              | Relevant      |
| ---                  | ---           |
| `aurora-data-api`    | ✅            |
| `aurora-data-api-pg` | ✅            |
| `better-sqlite3`     | ✅            |
| `cockroachdb`        | ✅            |
| `cordova`            | ✅            |
| `expo`               | ✅            |
| `mongodb`            | ❓            |
| `mysql`              | ✅            |
| `nativescript`       | ✅            |
| `oracle`             | ✅            |
| `postgres`           | ✅            |
| `react-native`       | ✅            |
| `sap`                | ❓            |
| `sqlite`             | ✅            |
| `sqlite-abstract`    | ✅            |
| `sqljs`              | ✅            |
| `sqlserver`          | ✅            |

### Are you willing to resolve this issue by submitting a Pull Request?

 - ✅ Yes, I have the time, and I know how to start.
 - ✖️ Yes, I have the time, but I don't know how to start. I would need guidance.
 - ✖️ No, I don’t have the time, but I can support (using donations) development.
 - ✖️ No, I don’t have the time and I’m okay to wait for the community / maintainers to resolve this issue.

Basically, I need someone to give green light for that.

**Comment by Ginden** [comment]
> Basically, I need someone to give green light for that.

@pleerock?
